### ISSUE TITLE
"stream is closed or unusable" caused by multithreading race-condition (pipe closed in another thread)

### ISSUE THREAD

**Issue by NHDaly** [JuliaLang/julia#47759]
We recently added a background-logging task to our server, but in some unit tests, we were seeing that very rarely the background task would fail with the following stack trace:
```julia
      nested task error: IOError: stream is closed or unusable
      Stacktrace:
        [1] check_open
          @ ./stream.jl:388 [inlined]
        [2] uv_write_async(s::Base.PipeEndpoint, p::Ptr{UInt8}, n::UInt64)
          @ Base ./stream.jl:1072
        [3] uv_write(s::Base.PipeEndpoint, p::Ptr{UInt8}, n::UInt64)
          @ Base ./stream.jl:1035
        [4] unsafe_write(s::Base.PipeEndpoint, p::Ptr{UInt8}, n::UInt64)
          @ Base ./stream.jl:1118
        [5] unsafe_write
          @ ./io.jl:683 [inlined]
        [6] write(s::Base.PipeEndpoint, a::Vector{UInt8})
          @ Base ./io.jl:706
        [7] handle_message(logger::Logging.ConsoleLogger, level::Base.CoreLogging.LogLevel, message::Any, _module::Any, group::Any, id::Any, filepath::Any, line::Any; kwargs::Base.Pairs{Symbol, V, Tuple{Vararg{Symbol, N}}, NamedTuple{names, T}} where {V, N, names, T<:Tuple{Vararg{Any, N}}})
          @ Logging /nix/store/60wrl8qkdrr2l6p9nbv8xff88a8igr0n-julia-1.8.2-patched/share/julia/stdlib/v1.8/Logging/src/ConsoleLogger.jl:178
[...]
```

We eventually tracked this down to a call to `@suppress_err` in the unit test, from Suppressors.jl, which temporarily redirects stderr to a pipe, and then puts it back and closes the pipe when the macro is done.

We believe this (rare) error is a race condition between the test's Task closing the pipe at the end of the macro, and the background logging task attempting to write to the pipe. That is, we think that one thread is redirecting stderr and closing the pipe inbetween these two calls:
checking that stderr isn't closed:
https://github.com/JuliaLang/julia/blob/6c0aa6dcbe10151501e4a1756876ea330734ab12/stdlib/Logging/src/ConsoleLogger.jl#L121-L123
and then writing to it here:
https://github.com/JuliaLang/julia/blob/6c0aa6dcbe10151501e4a1756876ea330734ab12/stdlib/Logging/src/ConsoleLogger.jl#L178

(Here's that whole section, for posterity):
https://github.com/JuliaLang/julia/blob/6c0aa6dcbe10151501e4a1756876ea330734ab12/stdlib/Logging/src/ConsoleLogger.jl#L121-L178

We're not sure what the right solution here should be.

It's possible that throwing an exception is already the right behavior: in our case, the exception stacktrace helped lead us to the issue!

On the other hand, philosophically, it seems that Julia takes the approach that an error during logging shouldn't throw an exception, so that you don't have to fear adding log messages. Additionally, it seems that after https://github.com/JuliaLang/julia/pull/40423, the logger is taking pains to gracefully fall back to stderr if the logger's stream has been closed. In this case, the stream was closed _after we checked_ (in a data race), so it's possible we can be more robust in our stream-handling if this safety really is our aim.

For example, we think that this could be fixed by adding a try/catch around the write, and falling back to stderr.

-------

Separately, we are going to file/fix the race condition in Suppressors.jl, probably by adding a delay in between redirecting stderr back to the original stderr, and closing the pipe.

CC: @dewilson, @vilterp

**Comment by vilterp** [comment]
It seems like if we're going to allow `stderr` and `stdout` to be mutated, we need to lock them somehow, e.g.

```julia
# declaration of stderr
stderr = …
stderr_lock = ReentrantLock()

# ...

function redirect_stderr(new_stderr)
  @lock stderr_lock
    stderr = new_stderr
  end
end

# ...

function handle_message(logger::ConsoleLogger)
  @lock stderr_lock
    stream = stderr
    # do a bunch of stuff
    write(stream, msg)
  end
end
```

But I'm not sure if that's feasible 🤔

**Comment by vtjnash** [comment]
Calling `redirect_stderr` is usually something libraries should not do internally, since it is global state, and thus inherently unreliably race-y. The functionality of `redirect_stdio` is commonly intended for things like application-startup or testing where it is necessary and desirable to intercept the normal behavior of things, but also is known that nothing else will be affected yet in the surrounding code.

**Comment by NHDaly** [comment]
> or testing where it is necessary and desirable to intercept the normal behavior of things, but also is known that nothing else will be affected yet in the surrounding code

Yeah, this was happening in test code, but it turned out that the code being tested had changed to spawn a background task that does some minor metrics logging, and that was causing this race condition.

I think we were just considering here whether the logger should be catching the possible errors from this race condition, so that logging cannot throw an exception. Currently, it can.

> On the other hand, philosophically, it seems that Julia takes the approach that an error during logging shouldn't throw an exception, so that you don't have to fear adding log messages.
